Adding multi-prompt support

ishmis opened the pull request.

dhil committed:
Abstract `fiber_{init,finalize}`

This patch abstracts away the fiber runtime initialisation and
finalisation procedures.

ishmis committed:
adding multi-prompt support

dhil commented on inc/fiber_prompt.h:
```diff
@@ -0,0 +1,46 @@
+#define export(NAME) __attribute__((export_name(NAME)))
+
+/** The abstract type of a prompt object **/
+typedef struct prompt* prompt_t;
```
```suggestion
typedef uint32_t prompt_t;
```
I suggest making it concrete.

dhil commented on inc/fiber_prompt.h:
```diff
@@ -0,0 +1,46 @@
+
+/** Yields control to the context of the provided prompt. This function must be
+   called from within a fiber context. **/
+export("fiber_yield") void* fiber_yield_to(void* arg, prompt_t prompt);
```
```suggestion
export("fiber_yield_to") void* fiber_yield_to(void* arg, prompt_t prompt);
```

dhil commented on src/fiber_prompt/wasmfx/imports.wat.pp:
```diff
@@ -0,0 +1,143 @@
+
+  (func $suspend_to (export "wasmfx_suspend_to") (param $arg i32) (param $name (ref $ht)) (result i32)
+    (suspend_to $ht $yield (local.get $arg) (local.get $name))
+    (drop)
```
It looks to me that you are dropping the name; though with the unsafe implementation of names it may continue to work, but this shouldn't really work. Instead the name should be stored into the handler name table.

dhil replied:
~~Alternatively you can update the table at `resume_with`.~~ On second thought, I don't think that will work.

ishmis replied:
yes this was temporary while I figured out when/how to put all names in a table

ishmis committed:
making prompt type concrete and fixing export naame for yield_to

ishmis committed:
ensuring active_fiber logic works with wasm tables

dhil commented on inc/fiber_prompt.h:
```diff
@@ -0,0 +1,48 @@
+
+/** Yields control to the context of the provided prompt. This function must be
+   called from within a fiber context. **/
+export("fiber_yield_to") void* fiber_yield_to(void* arg);
```
This isn't right. You definitively need to take a prompt as parameter here.
```suggestion
export("fiber_yield_to") void* fiber_yield_to(prompt_t prompt, void* arg);
```

dhil commented on src/fiber_prompt/wasmfx/imports.wat.pp:
```diff
@@ -0,0 +1,163 @@
+    (return)
+  )
+  
+  (func $suspend_to (export "wasmfx_suspend_to") (param $arg i32) (result i32)
+   (local $updated_name (ref null $ht))
+   (suspend_to $ht $yield (local.get $arg) (table.get $names (global.get $active_fiber)))
+   (local.set $updated_name)
+   (table.set $names (global.get $active_fiber) (local.get $updated_name))
+   (return)
+  )
```
I think this can be made more succinct.
```suggestion
  (func $suspend_to (export "wasmfx_suspend_to") (param $arg i32) (result i32)
   (suspend_to $ht $yield (local.get $arg) (table.get $names (global.get $active_fiber)))
   (table.set $names (global.get $active_fiber))
  )
```

dhil commented on src/fiber_prompt/wasmfx/wasmfx_impl.c:
```diff
@@ -0,0 +1,181 @@
+  return wasmfx_indexed_resume_with(table_index, arg, result);
+}
+
+void* fiber_yield_to(void* arg) {
```
Definitively need a prompt here, even if it isn't used. E.g.
```suggestion
void* fiber_yield_to(__attribute__((unused)) prompt_t p, void* arg) {
```

dhil commented on src/fiber_prompt/wasmfx/imports.wat.pp:
```diff
@@ -0,0 +1,163 @@
+    (global.set $sstack_current_ptrs_addr (local.get $sstack_current_ptrs_addr))
+  )
+  
+  (func $resume_trampoline (param $func_index i32) (param $arg i32) 
+                            (param $name (ref null $ht)) (result i32)
+    (table.set $names (global.get $active_fiber) (local.get $name))
+    (call_indirect $indirect_function_table (type $ft_client)
+      (local.get $arg)
+      (local.get $func_index)
+    )
+  )
+  (elem declare func $resume_trampoline)
+
+  (func $wasmfx_entry_trampoline (param $trampoline_index i32) (param $func_index i32) (param $arg i32) (param $name (ref null $ht)) (result i32)
+    (call_indirect $temp_table (type $ft2)
+      (local.get $arg)
+      (local.get $func_index)
+      (local.get $name)
+      (local.get $trampoline_index)
+    )
+  )
+  (elem declare func $wasmfx_entry_trampoline)
```
This doesn't look right. You should only need one trampoline.

ishmis committed:
working hello_world implementation

ishmis committed:
implementing asyncify with prompts

dhil commented on src/fiber_prompt/asyncify/asyncify_impl.c:
```diff
@@ -2,6 +2,7 @@
 
 #include <assert.h>
 #include <stdint.h>
+#include <stdio.h>
```
Stray import?

ishmis replied:
yep, will remove

dhil commented on src/fiber_prompt/asyncify/asyncify_impl.c:
```diff
@@ -140,14 +148,15 @@ __attribute__((noinline)) void fiber_free(fiber_t fiber) {
   if (active_fiber->state == YIELDING) {
     asyncify_stop_rewind();
     active_fiber->state = ACTIVE;
+    if (prompt != active_fiber->prompt) active_fiber->state = NAME_MISMATCH;
```
why are you checking the prompt name here?

dhil replied:
I'd expect that you check the prompt name after `stop_unwind`. And if it doesn't match, then initiate another unwind.

ishmis replied:
ah yes, that isn't necessary - will remove that. 

that's a good point, I think crafting a nested handler example will help test the scenario where that is necessary too - will get on that

ishmis committed:
generator example

dhil committed:
Prompt interface and bug fixes.

This patch is a stab at defining a prompt-oriented interface. I have
implemented a somewhat working version of it in Asyncify, though,
there are some caveats around forwarding. I think we need a trampoline
to make forwarding more robust.

dhil committed:
Add examples

ishmis committed:
correcting for updated asyncify multi_prompt backend

ishmis committed:
adding a simple parity checking program prompt example
